catalyst-gateway/bin/src/service/api/cardano/rbac/registrations_get/chain_info.rs

```diff
@@ -0,0 +1,134 @@
+use futures::{future::try_join, TryFutureExt, TryStreamExt};
+use rbac_registration::{cardano::cip509::Cip509, registration::cardano::RegistrationChain};
+
+use crate::{
+    db::index::{
+        queries::rbac::get_rbac_registrations::{Query, QueryParams},
+        session::CassandraSession,
+    },
+    settings::Settings,
+};
+
+/// A RBAC registration chain along with additional information.
+pub struct ChainInfo {
+    /// A RBAC registration chain.
+    pub chain: RegistrationChain,
+    /// The latest persistent transaction ID of the chain.
+    pub last_persistent_txn: Option<TransactionId>,
+    /// The latest volatile transaction ID of the chain.
+    pub last_volatile_txn: Option<TransactionId>,
+    /// A slot number of the latest persistent registration.
+    pub last_persistent_slot: Slot,
+}
+
+impl ChainInfo {
+    /// Creates a new `ChainInfo` instance.
+    pub(crate) async fn new(
+        persistent_session: &CassandraSession, volatile_session: &CassandraSession,
+        catalyst_id: &IdUri,
+    ) -> anyhow::Result<Option<Self>> {
+        let (persistent_registrations, volatile_registrations) = try_join(
+            indexed_registrations(persistent_session, catalyst_id),
+            indexed_registrations(volatile_session, catalyst_id),
+        )
+        .await?;
+        let network = Settings::cardano_network();
+
+        let (chain, last_persistent_slot) =
+            apply_registrations(network, None, persistent_registrations).await?;
+        let last_persistent_txn = chain.as_ref().map(RegistrationChain::current_tx_id_hash);
+
+        let (chain, _) = apply_registrations(network, chain, volatile_registrations).await?;
+        let mut last_volatile_txn = chain.as_ref().map(RegistrationChain::current_tx_id_hash);
+        if last_persistent_txn == last_volatile_txn {
+            last_volatile_txn = None;
+        }
+
+        Ok(chain.map(|chain| {
+            Self {
+                chain,
+                last_persistent_txn,
+                last_volatile_txn,
+                last_persistent_slot,
+            }
+        }))
+    }
+}
+
+/// Returns a sorted list of all registrations for the given Catalyst ID from the
+/// database.
+async fn indexed_registrations(
+    session: &CassandraSession, catalyst_id: &IdUri,
+) -> anyhow::Result<Vec<Query>> {
+    let mut result: Vec<_> = Query::execute(session, QueryParams {
+        catalyst_id: catalyst_id.clone().into(),
+    })
+    .and_then(|r| r.try_collect().map_err(Into::into))
+    .await?;
+
+    result.sort_by_key(|r| r.slot_no);
+    Ok(result)
+}
+
+/// Applies the given list of indexed registrations to the chain. If the given chain is
+/// `None` - new root will be possibly created.
+async fn apply_registrations(
+    network: Network, mut chain: Option<RegistrationChain>, indexed_registrations: Vec<Query>,
+) -> anyhow::Result<(Option<RegistrationChain>, Slot)> {
+    let mut last_slot = 0.into();
+
+    for reg in indexed_registrations {
```

Comment:
sort_by_key is sorting lowest to highest
when passing to this function it wont capture the latest?

Correct me if I'm wrong

Reply:
Unless I'm missing something this shouldn't affect selection of the latest _registration chain_ because we do that when we converting the given stake address to a Catalyst ID (in the `catalyst_id_from_stake` function).

Reply:
On second thought, you are probably right. I need to think how better to deal with it taking into account volatile registrations.